File: backend/app/schemas/member.py
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict


class MemberCreateRequest(BaseModel):
    name: str
    gender: str
    birth_date: date | None = None
    death_date: date | None = None
    is_alive: bool = True
    generation_no: int | None = None
    generation_name: str | None = None
    biography: str | None = None


class MemberUpdateRequest(BaseModel):
    name: str | None = None
    gender: str | None = None
    birth_date: date | None = None
    death_date: date | None = None
    is_alive: bool | None = None
    generation_no: int | None = None
    generation_name: str | None = None
    biography: str | None = None


class MemberListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    member_id: int
    tree_id: int
    name: str
    gender: str
    birth_date: date | None = None
    death_date: date | None = None
    is_alive: bool
    generation_no: int | None = None
    generation_name: str | None = None


class MemberDetailResponse(MemberListItem):
    biography: str | None = None
    created_at: datetime
    updated_at: datetime


class PaginatedMemberResponse(BaseModel):
    items: list[MemberListItem]
    total: int
    page: int
    page_size: int


class MemberIdRangeResponse(BaseModel):
    min_member_id: int | None = None
    max_member_id: int | None = None
    total: int

File: frontend/src/api/member.ts
import client from "./client";

export interface MemberListItem {
  member_id: number;
  tree_id: number;
  name: string;
  gender: "male" | "female" | "unknown";
  birth_date: string | null;
  death_date: string | null;
  is_alive: boolean;
  generation_no: number | null;
  generation_name: string | null;
}

export interface MemberDetailResponse extends MemberListItem {
  biography: string | null;
  created_at: string;
  updated_at: string;
}

export interface PaginatedMemberResponse {
  items: MemberListItem[];
  total: number;
  page: number;
  page_size: number;
}

export interface MemberCreatePayload {
  name: string;
  gender: "male" | "female" | "unknown";
  birth_date: string | null;
  death_date: string | null;
  is_alive: boolean;
  generation_no: number | null;
  generation_name: string | null;
  biography: string | null;
}

export interface MemberUpdatePayload {
  name?: string;
  gender?: "male" | "female" | "unknown";
  birth_date?: string | null;
  death_date?: string | null;
  is_alive?: boolean;
  generation_no?: number | null;
  generation_name?: string | null;
  biography?: string | null;
}

export function fetchMembers(treeId: number) {
  return client.get<PaginatedMemberResponse>(`/family-trees/${treeId}/members`);
}

export function fetchMemberDetail(treeId: number, memberId: number) {
  return client.get<MemberDetailResponse>(`/family-trees/${treeId}/members/${memberId}`);
}

export function createMember(treeId: number, payload: MemberCreatePayload) {
  return client.post<MemberDetailResponse>(`/family-trees/${treeId}/members`, payload);
}

export function updateMember(treeId: number, memberId: number, payload: MemberUpdatePayload) {
  return client.patch<MemberDetailResponse>(`/family-trees/${treeId}/members/${memberId}`, payload);
}

export function deleteMember(treeId: number, memberId: number) {
  return client.delete<{ message: string }>(`/family-trees/${treeId}/members/${memberId}`);
}
